### PR TITLE
Enable specifying exceptions in error macros

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This change enables libcudf APIs throwing exceptions other than `cudf::logic_error`. This code was adapted from rmm. This contributes to #10200.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
rerun tests

**Review by PointKernel — APPROVED** [review]

**Comment by vyasr** [comment]
> Only comment here would be to add a test or two.

@nvdbaranec I was planning to use these macros in a follow-up PR porting the changes from #10637. Would that be sufficient, or do you want to see the tests in this PR itself? I figured it would be fine to fix any bugs with the enhanced feature in the follow-up when I make use of it.

**Comment by vyasr** [comment]
@nvdbaranec thoughts? The new functionality should get tested in the next PR so I'm inclined to leave tests as is in this PR.

File: cpp/include/cudf/utilities/error.hpp [hunk 1]
```diff
@@ -99,37 +100,80 @@ struct fatal_cuda_error : public cuda_error {
...
+ * ```
+ * // throws cudf::logic_error
+ * CUDF_EXPECTS(p != nullptr, "Unexpected null pointer");
  *
- * @param[in] cond Expression that evaluates to true or false
- * @param[in] reason String literal description of the reason that cond is
- * expected to be true
- * @throw cudf::logic_error if the condition evaluates to false.
+ * // throws std::runtime_error
+ * CUDF_EXPECTS(p != nullptr, "Unexpected nullptr", std::runtime_error);
```
Comment (wence-) [inline cpp/include/cudf/utilities/error.hpp:112]: Is there a reason to flip the argument order of this macro from the existing one in RMM (which is `RMM_EXPECTS(condition, error, message)`)? I would argue that consistency across the libraries is probably worth it.

File: cpp/include/cudf/utilities/error.hpp [hunk 2]
```diff
@@ -99,37 +100,80 @@ struct fatal_cuda_error : public cuda_error {
...
+  (__VA_ARGS__)
+
+/// @cond
+
+#define GET_CUDF_EXPECTS_MACRO(_1, _2, _3, NAME, ...) NAME
+
+#define CUDF_EXPECTS_3(_condition, _reason, _exception_type)                        \
+  do {                                                                              \
+    static_assert(std::is_base_of_v<std::exception, _exception_type>);              \
+    (!!(_condition)) ? static_cast<void>(0)                                         \
```
Comment (wence-) [inline cpp/include/cudf/utilities/error.hpp:134]: Are there any circumstances under which `!!expr ? true : false` is different from `expr ? true : false` ? I think not because in the former the `!` contextually converts the expression to bool, whereas in the latter, the use as the first operand to `?` contextually converts to bool (IOW, forall a, `static_assert(!!a == bool(a), "Oh noes")` should succeed if conversion to bool is possible, I think)

**Review by wence- — COMMENTED** [review]
Nit: consistency with RMM for argument order?

File: cpp/include/cudf/utilities/error.hpp (hunk 2 above)
Comment (jrhemstad) [reply cpp/include/cudf/utilities/error.hpp:134]: The `!!` logic isn't necessary here. This is probably my fault from the original version I wrote.

The `!!` logic is useful when you're calling an API that takes an `int` that acts like a `bool` instead of a `bool`. e.g., some gcc built-ins work this way.

tl;dr this mechanism is useful in niche use cases, but not necessary here.

File: cpp/include/cudf/utilities/error.hpp (hunk 1 above)
Comment (vyasr) [reply cpp/include/cudf/utilities/error.hpp:112]: I didn't like the rmm order because it inserts an argument into the middle relative to the 2 argument form:
```
RMM_EXPECTS(_condition, _reason)
RMM_EXPECTS(_condition, _exception_type, _reason)  # Note that exception type comes _before_ reason
```
I don't know if there was a reason for that, @jrhemstad @harrism do you have any thoughts? If there's no particular reason I might instead prefer to modify the rmm macros for consistency with these new ones in cudf rather than try to match rmm.

File: cpp/include/cudf/utilities/error.hpp (hunk 2 above)
Comment (vyasr) [reply cpp/include/cudf/utilities/error.hpp:134]: Right, this is for doing the equivalent of `bool(x)` in Python AFAICT. I can remove it.

File: cpp/include/cudf/utilities/error.hpp (hunk 1 above)
Comment (jrhemstad) [reply cpp/include/cudf/utilities/error.hpp:112]: I've no problem with changing the order in RMM. I agree that this order is better.

Comment (vyasr) [reply cpp/include/cudf/utilities/error.hpp:112]: https://github.com/rapidsai/rmm/issues/1172

File: cpp/include/cudf/utilities/error.hpp [hunk 3]
```diff
@@ -99,37 +100,80 @@ struct fatal_cuda_error : public cuda_error {
...
+  (__VA_ARGS__)
+
+/// @cond
+
+#define GET_CUDF_EXPECTS_MACRO(_1, _2, _3, NAME, ...) NAME
+
+#define CUDF_EXPECTS_3(_condition, _reason, _exception_type)                    \
+  do {                                                                          \
+    static_assert(std::is_base_of_v<std::exception, _exception_type>);          \
+    _condition ? static_cast<void>(0)                                           \
```
Comment (vyasr) [inline cpp/include/cudf/utilities/error.hpp:134]:
```suggestion
    (_condition) ? static_cast<void>(0)                                         \
```

File: cpp/include/cudf/utilities/error.hpp (hunk 1 above)
Comment (vyasr) [reply cpp/include/cudf/utilities/error.hpp:112]: No, the message isn't optional, the exception type is. You either do `RMM_EXPECTS(condition, message)` or you do `RMM_EXPECTS(condition, exception_type, message)` if you want a different exception than just an `rmm::logic_error`.

**Review by wence- — APPROVED** [review]
I think my approval has no power here, but looks good (except the clang-format issues, editing in the web-interface is not always better ;P)

**Comment by vyasr** [comment]
@gpucibot merge
